From 6450b6e320b978d8e49c64aeb56f227bb60f5bfa Mon Sep 17 00:00:00 2001
From: Jan Moeller <Jan.Moeller@governikus.de>
Date: Mon, 1 Sep 2025 13:15:14 +0200
Subject: a11y Android: Inform TalkBack about used language

This is helpful if the app's language is different from the OS language.
Without this change english texts are read with a german accent if
the OS language is set to German and the app language to English.

Users will have to set QLocale::setDefault to their desired app
language in order for this behavior to work.

Pick-to: 6.8 6.10
Change-Id: Id4bf0280469229c202ab3cf2480903212d351c2b
---
 .../qt/android/QtAccessibilityDelegate.java   | 21 +++++++++++++++++--
 .../qt/android/QtNativeAccessibility.java     |  1 +
 .../android/androidjniaccessibility.cpp       |  7 +++++++
 3 files changed, 27 insertions(+), 2 deletions(-)

diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
index bc19882840d74ed2b8a6cb40323ed73508b1a6ce..7f64928b6cd14517ecf60b63ec2a29eaec2fce4b 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
@@ -10,6 +10,10 @@ import android.os.Build;
 import android.os.Bundle;
 import android.system.Os;
 import android.text.TextUtils;
+import android.text.SpannableString;
+import android.text.Spanned;
+import android.text.style.LocaleSpan;
+import java.util.Locale;
 import android.util.Log;
 import android.view.MotionEvent;
 import android.view.View;
@@ -166,6 +170,15 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         return true;
     }
 
+    SpannableString addLocaleSpan(String value)
+    {
+        SpannableString localeValue = new SpannableString(value);
+        LocaleSpan localeSpan =
+                new LocaleSpan(Locale.forLanguageTag(QtNativeAccessibility.appLanguage()));
+        localeValue.setSpan(localeSpan, 0, localeValue.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
+        return localeValue;
+    }
+
     void notifyScrolledEvent(int viewId)
     {
         QtNative.runAction(() -> sendEventForVirtualViewId(viewId,
@@ -248,7 +261,7 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
             event.setEnabled(true);
             event.setClassName(getNodeForVirtualViewId(viewId).getClassName());
 
-            event.setContentDescription(value);
+            event.setContentDescription(addLocaleSpan(value));
 
             if (event.getText().isEmpty() && TextUtils.isEmpty(event.getContentDescription())) {
                 Log.w(TAG, "No value to announce for " + event.getClassName());
@@ -355,7 +368,9 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         event.setEnabled(true);
         event.setClassName(getNodeForVirtualViewId(virtualViewId).getClassName());
 
-        event.setContentDescription(QtNativeAccessibility.descriptionForAccessibleObject(virtualViewId));
+        String description = QtNativeAccessibility.descriptionForAccessibleObject(virtualViewId);
+        event.setContentDescription(addLocaleSpan(description));
+
         if (event.getText().isEmpty() && TextUtils.isEmpty(event.getContentDescription()))
             Log.w(TAG, "AccessibilityEvent with empty description");
 
@@ -458,6 +473,8 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         // set only if valid, otherwise we return a node that is invalid and will crash when accessed
         node.setSource(m_view, virtualViewId);
 
+        node.setContentDescription(addLocaleSpan(node.getContentDescription().toString()));
+
         if (TextUtils.isEmpty(node.getText()) && TextUtils.isEmpty(node.getContentDescription()))
             Log.w(TAG, "AccessibilityNodeInfo with empty contentDescription: " + virtualViewId);
 
diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
index e63de6f79756198f1f3de1e4fc894c6a4748d2c4..cfef859ab05d90a38de53c0a53a8dc2d4a1e4967 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
@@ -12,6 +12,7 @@ class QtNativeAccessibility
     static native int[] childIdListForAccessibleObject(int objectId);
     static native int parentId(int objectId);
     static native String descriptionForAccessibleObject(int objectId);
+    static native String appLanguage();
     static native Rect screenRect(int objectId);
     static native int hitTest(float x, float y);
     static native boolean clickAction(int objectId);
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index a88bfc00eaf57f995465023d55560410c8f61495..8de9758bb0f7a1f7c27c3cd1081a2074f8d10738 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -19,6 +19,7 @@
 #include <QtCore/QObject>
 #include <QtCore/qpointer.h>
 #include <QtCore/qvarlengtharray.h>
+#include <QLocale>
 
 static const char m_qtTag[] = "Qt A11Y";
 
@@ -638,6 +639,11 @@ namespace QtAndroidAccessibility
         return env->NewString((jchar*) desc.constData(), (jsize) desc.size());
     }
 
+    static jstring appLanguage(JNIEnv *env, jobject /*thiz*/)
+    {
+        const QString localeName = QLocale().bcp47Name();
+        return env->NewString((jchar *)localeName.constData(), (jsize)localeName.size());
+    }
 
     struct NodeInfo
     {
@@ -742,6 +748,7 @@ namespace QtAndroidAccessibility
         {"childIdListForAccessibleObject", "(I)[I", (jintArray)childIdListForAccessibleObject},
         {"parentId", "(I)I", (void*)parentId},
         {"descriptionForAccessibleObject", "(I)Ljava/lang/String;", (jstring)descriptionForAccessibleObject},
+        {"appLanguage", "()Ljava/lang/String;", (jstring)appLanguage},
         {"screenRect", "(I)Landroid/graphics/Rect;", (jobject)screenRect},
         {"hitTest", "(FF)I", (void*)hitTest},
         {"populateNode", "(ILandroid/view/accessibility/AccessibilityNodeInfo;)Z", (void*)populateNode},
